File: leetcode/medium/518. Coin Change 2.py
# problem : https://leetcode.com/problems/coin-change-2/
# time complexity : O(NM)
# algorithm : DP
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        if amount == 0 and len(coins) == 0:
            return 1
        dp = [[0] * (amount + 1)] * (len(coins) + 1)

        for coinInd in range(len(coins)):
            coin = coins[coinInd]
            prev = dp[coinInd]
            now = dp[coinInd + 1]
            now[0] = 1
            for ind in range(coin, amount + 1):
                now[ind] = now[ind-coin] + prev[ind]

        return dp[len(coins)][amount]

File: leetcode/medium/518. Coin Change 2.cpp
/**
 * problem : https://leetcode.com/problems/coin-change-2/
 * time complexity : O(NM). N = amount, M = |coins|
 * algorithm : DP
 */
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        vector<int> dp(amount + 1, 0);
        dp[0] = 1;
        for(int i=1; i<=coins.size(); i++){
            int coin = coins[i-1];
            for(int j=coin; j<=amount; j++){
                dp[j] += dp[j-coin];
            }
        }
        return dp[amount];
    }
};
